### PR TITLE
AdminGenreController done without errors

### PR DESCRIPTION
**Pull request by trevorhuis-smoothstack**
It seems as though Spring Boot will allow us to create a global error handler. Maybe this can be accomplished in the orchestrator. we can ask akbar later so i am going to just put up the barebones controllers for now. Let me know it looks

### CODE REVIEW

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (sean-darsie) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: The resonse code should not be "created" here because this is a get method right.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (sean-darsie) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: surround with try catch

**Review by sean-darsie — APPROVED** [review]
Need to add some more try catch blocks for error handling but otherwise it is good.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (trevorhuis-smoothstack) [reply src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: good catch!

**Review by justin-obrien-smoothstack — CHANGES_REQUESTED** [review]
I don't think our controller methods should be throwing any exceptions; they should be handling any exceptions that get thrown to them.

In the event of an SQLException, it seems to me the best status code is 500 (Internal Server Error). If an SQLException was thrown it basically means one of our database access methods failed, and the client's request may have been fine.

I'm pretty sure we're supposed to make each method return a status code to the client, probably via a ResponseEntity object.

I think each method (other than create) should return 404 if the client requests something that doesn't exist in the database.

**Comment by justin-obrien-smoothstack** [comment]
Another thought: It seems that it's best to have the service classes start throwing exceptions instead of catching them, so the service controllers can catch the exceptions and return the appropriate status codes accordingly. Currently the service classes print to the terminal in response to exceptions, but that won't do much good within our current paradigm.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: Here, the status should be CREATED instead of OK.

**Review by justin-obrien-smoothstack — CHANGES_REQUESTED** [review]
If a service method uses a boolean to keep track of whether its operation succeeded, the success variable should be set to true before the method returns, so the finally block can take the correct action based on it. Your two read methods are capable of returning results of a successful operation before success is set to true. However, I don't think this will actually cause any problems in terms of functionality, because the read methods don't need the success boolean; commit and rollback are only needed when database entries are modified. It would be good to clean those methods up, but I'm pretty sure they'll run fine as-is so they're not necessarily among the changes I'm requesting.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: Have you considered setting the status to No Content here? I was thinking about it, and I kind of feel like it makes sense: The client requested all records from a given table. We have the table but it has no records, so we basically tell them we successfully processed the request and the result is no content. Not Found, on the other hand, would seem to suggest the table doesn't exist. Just food for thought.

**Comment by justin-obrien-smoothstack** [comment]
Another note: The database schema does allow for genre name to be null, so we don't strictly have to check for it, but if we want to prevent clients from making genre names null then I don't see anything wrong with that.

One thing the database won't allow, though, is a genre name longer than 45 characters. Now that I think of it, we should certainly check for that.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: We shouldn't check for a null name here. It could be that the user wants to delete a genre whose name is null (which the database schema allows). All that matters is whether the user has specified a valid genre ID. For that matter, it could be that the ID isn't null but also doesn't correspond to any genre in the database.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: Even if the ID is non-null,  it could still be invalid if it doesn't belong to any genre in the database.

**Review by sean-darsie — APPROVED** [review]
Approved, once you make the changes justin mentioned go ahead and merge

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: This will hit an exception if the genre isn't null but its name is. To avoid this, the second condition can be replaced with `(genre.getGenreName() != null && genre.getGenreName().length() > 45)`.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: Same as previous comment, regarding the possibility of the genre name being null.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java:104]: I've been using the OK status for successful updates. I'm not sure which one is more fitting, but we should decide on one for consistency throughout the application.

**Review by justin-obrien-smoothstack — CHANGES_REQUESTED** [review]
Everything looks good, except that we should make sure the genre name is non-null before calling length() on it.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (trevorhuis-smoothstack) [reply src/main/java/com/ss/training/lms/controller/AdminGenreController.java:104]: I was thinking about this too. Should a response code like UPDATED

**Comment by trevorhuis-smoothstack** [comment]
I fixed the two null checks. We can resolve the response code in the slack channel.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [reply src/main/java/com/ss/training/lms/controller/AdminGenreController.java:104]: Yeah, I was surprised to learn there's no such code.

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: It's still possible to check the length of a null string here.

**Review by justin-obrien-smoothstack — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/training/lms/controller/AdminGenreController.java
Comment (trevorhuis-smoothstack) [reply src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: geez my bad. I misunderstood your comment.

Comment (justin-obrien-smoothstack) [reply src/main/java/com/ss/training/lms/controller/AdminGenreController.java]: No worries! I could have been more specific.

**Review by justin-obrien-smoothstack — APPROVED** [review]

**Pull request merged**
